fix null ref exception for nuget pack projects which have a dependency on nupkg with a *.transform content file

rohit21agrawal opened the pull request:
fixes : https://github.com/NuGet/Home/issues/3160

Since the web.config.transform trying to be read is from the nupkg, we need to read it while the ZipArchive has not been disposed and store it in a MemoryStream to be retrieved later.

CC: @emgarten @alpaix @joelverhagen @drewgil @rrelyea @jainaashish

Commit:
fix null ref exception for nuget pack projects which have a dependency on nupkg with a *.transform content file

alpaix commented on src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:
```diff
@@ -1066,7 +1066,20 @@ private static string RemoveExtension(string path)
-            return groups.SelectMany(g => g.Items).Where(IsTransformFile).Select(f => new Packaging.PhysicalPackageFile() { TargetPath = f });
+            return groups.SelectMany(g => g.Items).Where(IsTransformFile).Select(f =>
+            {
+                Packaging.PhysicalPackageFile file = new Packaging.PhysicalPackageFile()
```
nit: var

alpaix commented on src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PhysicalPackageFile.cs:
```diff
@@ -22,6 +22,8 @@ internal PhysicalPackageFile(Func<Stream> streamFactory)
             _streamFactory = streamFactory;
         }
 
+        public MemoryStream MemoryStream { get; set; } 
```
this class is immutable. Consider changing this attribute to read-only along with providing relevant constructor.

alpaix commented on src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:
```diff
@@ -1066,7 +1066,20 @@ private static string RemoveExtension(string path)
+                    MemoryStream = new MemoryStream()
+                };
+
+                XElement element = XElement.Load(package.GetStream(f));
```
`var` here

alpaix commented on src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:
```diff
@@ -1066,7 +1066,20 @@ private static string RemoveExtension(string path)
+
+                XElement element = XElement.Load(package.GetStream(f));
+                element.Save(file.MemoryStream);
+                file.MemoryStream.Seek(0, SeekOrigin.Begin);
```
I think it'd better if `file` would get created after the memory stream is fully initialized.

alpaix commented:
Also consider adding a functional test for that scenario.

emgarten commented:
:+1: on adding a test

rohit21agrawal commented:
@emgarten @alpaix Have made changes as per your suggestions and added a functional test as well.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PhysicalPackageFile.cs:
```diff
@@ -17,11 +17,18 @@ public PhysicalPackageFile()
             _streamFactory = streamFactory;
         }
 
+        public MemoryStream MemoryStream { get; private set; } 
```
why not make it private?

rohit21agrawal replied:
fixed

jainaashish commented:
:shipit:

emgarten commented:
:shipit:

rohit21agrawal commented:
@alpaix @emgarten I have removed the end to end test, and added another unit test, as the end-to-end test didn't use the latest built version of nuget.exe.

alpaix commented:
looks good! :shipit:
